src/include/miopen/inline_vector.hpp

```diff
@@ -0,0 +1,261 @@
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef GUARD_MIOPEN_INLINE_VECTOR_HPP
+#define GUARD_MIOPEN_INLINE_VECTOR_HPP
+
+#include <array>
+#include <miopen/config.h>
+#include <miopen/errors.hpp>
+
+namespace miopen {
+
+template <typename T, std::size_t N>
+class InlineVector
+{
+public:
+    using storage_type           = std::array<T, N>;
+    using value_type             = typename storage_type::value_type;
+    using size_type              = typename storage_type::size_type;
+    using reference              = typename storage_type::reference;
+    using const_reference        = typename storage_type::const_reference;
+    using pointer                = typename storage_type::pointer;
+    using const_pointer          = typename storage_type::const_pointer;
+    using iterator               = typename storage_type::iterator;
+    using const_iterator         = typename storage_type::const_iterator;
+    using reverse_iterator       = typename storage_type::reverse_iterator;
+    using const_reverse_iterator = typename storage_type::const_reverse_iterator;
+    static_assert(std::is_scalar_v<T>, "InlineVector currently supports scalar type only");
+
+    // Default constructor
+    InlineVector() = default;
+
+    // Copy and move constructor
+    InlineVector(const InlineVector& inline_vec)     = default;
+    InlineVector(InlineVector&& inline_vec) noexcept = default;
+
+    InlineVector(std::initializer_list<T> data) : real_size(data.size())
+    {
+        if(real_size > N)
+        {
+            MIOPEN_THROW("Input data size is bigger than InlineVector's capacity");
+        }
+
+        std::copy(data.begin(), data.end(), storage.begin());
+    }
+
+    template <typename _InputIterator>
```

Comment:
_InputIterator is reserved by the standard (underscore + uppercase), you can just name it InputIterator or Iterator